ann_benchmarks/plotting/metrics.py

```diff
@@ -1,31 +1,70 @@
 from __future__ import absolute_import
+import numpy as np
 
-def knn(dataset_distances, run_distances, count, epsilon=1e-10):
-    total = len(run_distances) * count
-    actual = 0
-    for true_distances, found_distances in zip(dataset_distances, run_distances):
-        within = [d for d in found_distances[:count] if d <= true_distances[count - 1] + epsilon]
-        actual += len(within)
-    return float(actual) / float(total)
+def knn_threshold(data, count, epsilon):
+    return data[count - 1] + epsilon
+
+def epsilon_threshold(data, count, epsilon):
+    return data[count - 1] * (1 + epsilon)
 
-def epsilon(dataset_distances, run_distances, count, epsilon=0.01):
+def get_recall_values(dataset_distances, run_distances, count, threshold, epsilon=1e-3):
     total = len(run_distances) * count
-    actual = 0
-    for true_distances, found_distances in zip(dataset_distances, run_distances):
-        within = [d for d in found_distances[:count] if d <= true_distances[count - 1] * (1 + epsilon)]
-        actual += len(within)
-    return float(actual) / float(total)
+    recalls = np.zeros(len(run_distances))
+    for i in range(len(run_distances)):
+        t = threshold(dataset_distances[i], count, epsilon)
+        actual = 0
+        for j in range(min(count, len(run_distances[i]))):
```

Comment:
`for j, d in enumerate(run_distances[i][:count]):` :)

Reply:
Thanks! Much better. Changed it in cdda15b. (Also removed the break statement because it assumes that algorithms sort their answers to queries.)